pkg/chains/formats/slsa/v2alpha2/internal/taskrun/taskrun.go

```diff
@@ -0,0 +1,112 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package taskrun
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	slsa "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v1"
+	"github.com/tektoncd/chains/pkg/chains/formats/slsa/extract"
+	resolveddependencies "github.com/tektoncd/chains/pkg/chains/formats/slsa/v2alpha2/internal/resolved_dependencies"
+	"github.com/tektoncd/chains/pkg/chains/objects"
+)
+
+const taskRunResults = "taskRunResults/%s"
+
+// GenerateAttestation generates a provenance statement with SLSA v1.0 predicate for a task run.
+func GenerateAttestation(ctx context.Context, builderID string, tro *objects.TaskRunObject) (interface{}, error) {
+	rd, err := resolveddependencies.TaskRun(ctx, tro)
+	if err != nil {
+		return nil, err
+	}
+	bp, err := byProducts(tro)
+	if err != nil {
+		return nil, err
+	}
+	att := intoto.ProvenanceStatementSLSA1{
+		StatementHeader: intoto.StatementHeader{
+			Type:          intoto.StatementInTotoV01,
+			PredicateType: slsa.PredicateSLSAProvenance,
+			Subject:       extract.SubjectDigests(ctx, tro),
+		},
+		Predicate: slsa.ProvenancePredicate{
+			BuildDefinition: slsa.ProvenanceBuildDefinition{
+				BuildType:            "https://tekton.dev/chains/v2/slsa",
+				ExternalParameters:   externalParameters(tro),
+				InternalParameters:   internalParameters(tro),
+				ResolvedDependencies: rd,
+			},
+			RunDetails: slsa.ProvenanceRunDetails{
+				Builder: slsa.Builder{
+					ID: builderID,
+				},
+				BuildMetadata: metadata(tro),
+				Byproducts:    bp,
+			},
+		},
+	}
+	return att, nil
+}
+
+func metadata(tro *objects.TaskRunObject) slsa.BuildMetadata {
+	m := slsa.BuildMetadata{
+		InvocationID: string(tro.ObjectMeta.UID),
+	}
+	if tro.Status.StartTime != nil {
+		utc := tro.Status.StartTime.Time.UTC()
+		m.StartedOn = &utc
+	}
+	if tro.Status.CompletionTime != nil {
+		utc := tro.Status.CompletionTime.Time.UTC()
+		m.FinishedOn = &utc
+	}
+	return m
+}
+
+// internalParameters adds the tekton feature flags that were enabled
+// for the taskrun.
+func internalParameters(tro *objects.TaskRunObject) map[string]any {
+	internalParams := make(map[string]any)
+	if tro.Status.Provenance != nil && tro.Status.Provenance.FeatureFlags != nil {
+		internalParams["tekton-pipelines-feature-flags"] = *tro.Status.Provenance.FeatureFlags
+	}
+	return internalParams
+}
+
+// externalParameters adds the task run spec
+func externalParameters(tro *objects.TaskRunObject) map[string]any {
+	externalParams := make(map[string]any)
+	externalParams["runSpec"] = tro.Spec
+	return externalParams
+}
+
+// byProducts contains the taskRunResults
+func byProducts(tro *objects.TaskRunObject) ([]slsa.ResourceDescriptor, error) {
```

Comment:
NIT: Consider de-duplicating these functions (`metadata`, `internalParameters`, `exeternalParameters`, `byproducts`) between taskrun.go and pipelinerun.go. The `TektonObject` interface can help here.

Reply:
I looked into this. It's a little tricky since the underlying struct of pipelineRunSpec vs taskRunSpec is a little different. We could clean it up but it will end up being a significant refactoring. Could we do it in a follow up PR if its something we absolutely want to do.

Reply:
We probably would have to end up declaring the complete type like with results : `GetResults`
https://github.com/tektoncd/chains/blob/cbd9495f75df3d7e821845631fddd2e1e7482f56/pkg/chains/objects/objects.go#L44-L48

Reply:
Yeah. Definitely not a blocker here. Just wanted to hear your thoughts on it :)